clients/src/main/java/org/apache/kafka/common/requests/UpdateMetadataRequest.java

```diff
@@ -16,192 +16,90 @@
-            LIVE_BROKERS_V2);
-
-
-    private static final Schema UPDATE_METADATA_REQUEST_V3 = new Schema(
-            CONTROLLER_ID,
-            CONTROLLER_EPOCH,
-            PARTITION_STATES_V0,
-            LIVE_BROKERS_V3);
-
-    // UPDATE_METADATA_REQUEST_V4 added a per-partition offline_replicas field. This field specifies the list of replicas that are offline.
-    private static final Schema UPDATE_METADATA_REQUEST_V4 = new Schema(
-            CONTROLLER_ID,
-            CONTROLLER_EPOCH,
-            PARTITION_STATES_V4,
-            LIVE_BROKERS_V3);
-
-    // UPDATE_METADATA_REQUEST_V5 added a broker_epoch Field. This field specifies the generation of the broker across
-    // bounces. It also normalizes partitions under each topic.
-    private static final Schema UPDATE_METADATA_REQUEST_V5 = new Schema(
-            CONTROLLER_ID,
-            CONTROLLER_EPOCH,
-            BROKER_EPOCH,
-            TOPIC_STATES_V5,
-            LIVE_BROKERS_V3);
-
-    public static Schema[] schemaVersions() {
-        return new Schema[] {UPDATE_METADATA_REQUEST_V0, UPDATE_METADATA_REQUEST_V1, UPDATE_METADATA_REQUEST_V2,
-            UPDATE_METADATA_REQUEST_V3, UPDATE_METADATA_REQUEST_V4, UPDATE_METADATA_REQUEST_V5};
-    }
 
     public static class Builder extends AbstractControlRequest.Builder<UpdateMetadataRequest> {
-        private final Map<TopicPartition, PartitionState> partitionStates;
-        private final Set<Broker> liveBrokers;
+        private final List<UpdateMetadataPartitionState> partitionStates;
+        private final List<UpdateMetadataBroker> liveBrokers;
 
         public Builder(short version, int controllerId, int controllerEpoch, long brokerEpoch,
-                       Map<TopicPartition, PartitionState> partitionStates, Set<Broker> liveBrokers) {
+                       List<UpdateMetadataPartitionState> partitionStates, List<UpdateMetadataBroker> liveBrokers) {
             super(ApiKeys.UPDATE_METADATA, version, controllerId, controllerEpoch, brokerEpoch);
             this.partitionStates = partitionStates;
             this.liveBrokers = liveBrokers;
         }
 
         @Override
         public UpdateMetadataRequest build(short version) {
-            if (version == 0) {
-                for (Broker broker : liveBrokers) {
-                    if (broker.endPoints.size() != 1 || broker.endPoints.get(0).securityProtocol != SecurityProtocol.PLAINTEXT) {
-                        throw new UnsupportedVersionException("UpdateMetadataRequest v0 only handles PLAINTEXT endpoints");
+            if (version < 3) {
+                for (UpdateMetadataBroker broker : liveBrokers) {
+                    if (version == 0) {
+                        if (broker.endpoints().size() != 1 || broker.endpoints().get(0).securityProtocol() != SecurityProtocol.PLAINTEXT.id) {
+                            throw new UnsupportedVersionException("UpdateMetadataRequest v0 only handles PLAINTEXT endpoints");
```

Comment:
Should we have a separate `UnsupportedVersionException` for multiple endpoints?  For example, we could have multiple plaintext endpoints.